File: 分治/23.合并K个升序链表.py
class Solution:
    # approach1: heap
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not any(lists):  # any(array) 当数组中有一个值为True时返回True
            return
        # 将所有非空列表的 值,列表索引,列表头节点 加入heap
        heap = [(l.val, i, l) for i, l in enumerate(lists) if l]
        heapq.heapify(heap)

        dummy = curr = ListNode()  # 新建一个dummy Node和一个指针
        while heap:  # 在最后两次遍历之前， heap中项目的个数都为k个？ 就像heap在过滤k条链表
            _, i, node = heapq.heappop(heap)
            curr.next = node
            curr = curr.next
            # 如果这个节点有后续节点将其加入优先队列
            if node.next:
                heapq.heappush(heap, (node.next.val, i, node.next))

        return dummy.next

File: 分治/23.合并K个升序链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
  if (lists.length === 0) return null;
  if (lists.length === 1) return lists[0];

  function mergeListsRange(start, end) {
    // 基本情况,只有一条链表
    if (start === end) return lists[start];
    const m = (start + end) >> 1;
    return merge(mergeListsRange(start, m), mergeListsRange(m + 1, end));
  }

  return mergeListsRange(0, lists.length - 1);
};

function merge(l1, l2) {
  const dummy = new ListNode();
  let cur = dummy;
  while (l1 && l2) {
    if (l1.val < l2.val) {
      cur.next = l1;
      l1 = l1.next;
    } else {
      cur.next = l2;
      l2 = l2.next;
    }
    cur = cur.next;
  }
  cur.next = l1 || l2;
  return dummy.next;
}
